### ISSUE TITLE
Oneapi dosen't work on nvidia gpu

### ISSUE THREAD

**Issue by Alkeryn** [intel/llvm#879]
As there is no oneapi repo yet i'm posting the issue here
anyway i have some sycl code that works on my intel integrated GPU
but it does not work on my nvidia gpu, here is an example :
![image](https://user-images.githubusercontent.com/11599075/69731766-bdde9100-112a-11ea-8a66-c7a0c0aa09c6.png)
the compiled code is : 
```cpp
#include <iostream>
#include <CL/sycl.hpp>
#include <experimental/random>

using namespace cl::sycl;

class vector_addition;

class myselector : public device_selector {
    public:
	int operator()(const device &dev) const override{
	    int rating = 0;
	    std::cout<<dev.get_info<info::device::name>()<<std::endl;
	    /* if (dev.is_gpu()) rating = 1; */
	    /* if (dev.is_cpu()) rating = -100; */
	    if(dev.get_info<info::device::name>().find("GTX") != std::string::npos) rating=1;
	    /* if(dev.get_info<info::device::name>().find("NEO") != std::string::npos) rating=1; */
	    return rating;
	};
};


int main(int, char**) {
   cl::sycl::float4 a = { 1.0, 2.0, 3.0, 4.0 };
   cl::sycl::float4 b = { 4.0, 3.0, 2.0, 1.0 };
   cl::sycl::float4 c = { 0.0, 0.0, 0.0, 0.0 };

   /* cl::sycl::default_selector device_selector; */
   myselector device_selector;

   cl::sycl::queue queue(device_selector);
   std::cout << "Running on "
             << queue.get_device().get_info<cl::sycl::info::device::name>()
             << "\n";
   {
      cl::sycl::buffer<cl::sycl::float4, 1> a_sycl(&a, cl::sycl::range<1>(1));
      cl::sycl::buffer<cl::sycl::float4, 1> b_sycl(&b, cl::sycl::range<1>(1));
      cl::sycl::buffer<cl::sycl::float4, 1> c_sycl(&c, cl::sycl::range<1>(1));

      queue.submit([&] (cl::sycl::handler& cgh) {
         auto a_acc = a_sycl.get_access<cl::sycl::access::mode::read>(cgh);
         auto b_acc = b_sycl.get_access<cl::sycl::access::mode::read>(cgh);
         auto c_acc = c_sycl.get_access<cl::sycl::access::mode::read_write>(cgh);

         cgh.single_task<class vector_addition>([=] () {
         c_acc[0] = a_acc[0] + b_acc[0];
         });
      });
   }
   std::cout << "  A { " << a.x() << ", " << a.y() << ", " << a.z() << ", " << a.w() << " }\n"
        << "+ B { " << b.x() << ", " << b.y() << ", " << b.z() << ", " << b.w() << " }\n"
        << "------------------\n"
        << "= C { " << c.x() << ", " << c.y() << ", " << c.z() << ", " << c.w() << " }"
        << std::endl;

   return 0;
}
```

**Comment by Ruyk** [comment]
Codeplay is working on OneAPI CUDA support, but it is currently not available on the main DPC++ compiler. We will be open-sourcing our work early next year, and that will provide some early experimental support for CUDA 10.1 on platforms with at least SM 3.0 capabilities using DPC++.
Keep watching this space :-)

**Comment by Alkeryn** [comment]
Thanks for the information, i'll be using hipsycl in the meanwhile ^

**Comment by keryell** [comment]
You could also ask Nvidia to implement it due to popular demand... ;-)

**Comment by Alkeryn** [comment]
@keryell I think it is easier to just avoid depending on nvidia as how annoying they can be with following standards tho lol
Also it wouldn't make a lot of sense for them as they are still trying to push people to use shitty cuda instead.

A thing cuda/hip has i wish opencl and sycl had tho is dynamic memory allocation inside a gpu kernel.

**Comment by keryell** [comment]
You can implement dynamic allocation in SYCL as a library if you have atomics & USM extensions.

**Comment by jeffhammond** [comment]
I use [ComputeCpp](https://codeplay.com/products/computesuite/computecpp) PTX support on Volta.  ComputeCpp is free as in beer.  As noted above, CodePlay is working an open-source SYCL compiler for NVIDIA.

**Comment by jeffhammond** [comment]
@Alkeryn You should create a new/separate issue on the topic of dynamic memory allocation inside a kernel.  This is a known issue under investigation, but having external users such as yourself request it helps to prioritize it.

You probably know this, but CUDA/HIP allocation inside of kernels is a hack that you can implement yourself pretty easily in SYCL.  As described in [CUDA docs](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#dynamic-global-memory-allocation-and-operations), there is a preallocated memory pool that defaults to 8 MB to support `malloc` and `new`.  I haven't tried C++ PMR with SYCL buffers yet, but any suballocator implementation should give you `malloc` in SYCL with the same design limitations of CUDA's implementation.

**Comment by Alkeryn** [comment]
@jeffhammond Yeah that's what i do generally, i just give N memory to the gpu beforehand and implement my own "malloc" for use in kernel
still a true malloc would be nice, also i'd love to be able to use the stl containers in a kernel or have an stl gpu implementation (as that would make porting code easier)
but i guess i can do without ^

**Comment by jeffhammond** [comment]
@Alkeryn I agree that your feature request is valid and we should do it, I just wanted to make sure you weren't waiting on this feature before getting started.

**Comment by AlexeySachkov** [comment]
@Alkeryn, CUDA support was merged about a month ago and it is included into [2020-03 release](https://github.com/intel/llvm/releases/tag/2020-03)

Have you tried it? Can we close this issue?
